File: 0234-palindrome-linked-list/0234-palindrome-linked-list.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {boolean}
 */

const reverseLinkedList = (head) => {
    let prev = null;
    
    while (head !== null) {
        let tempHeadNext = head.next;
        head.next = prev;
        prev = head;
        head = tempHeadNext;
    }
    
    return prev;
};

var isPalindrome = function(head) {
    let fast = head;
    let slow = head;
    let startPointer = head;
    let length = 0;
    
    while (fast && fast.next) {
        fast = fast.next.next;
        slow = slow.next;
        length++;
    }
    
    let mid = reverseLinkedList(slow);
    
    while (length) {
        length--;
        if (mid.val !== startPointer.val) return false;
        mid = mid.next;
        startPointer = startPointer.next;
    }
    
    return true;
};

File: 0234-palindrome-linked-list/0234-palindrome-linked-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
def reverseLinkedList(head):
    prev = None
    currNode = head
    nextNode = currNode.next
    while (nextNode != None):
        currNode.next = prev
        prev = currNode
        currNode = nextNode
        nextNode = nextNode.next

    currNode.next = prev
    return currNode

class Solution:        
    def isPalindrome(self, head: Optional[ListNode]) -> bool:
        fast = head
        slow = head
        while (fast and fast.next):
            fast = fast.next.next
            slow = slow.next
        
        secondHalfReversed = reverseLinkedList(slow)
        while (head.next != None):
            if (head.val != secondHalfReversed.val):
                return False
            head = head.next
            secondHalfReversed = secondHalfReversed.next 
        return True
